readme.md
[![Build Status](https://travis-ci.org/Tennu/tennu.png?branch=master)](https://travis-ci.org/Tennu/tennu)

Tennu is an IRC bot framework written in Node.js

[![NPM](https://nodei.co/npm/tennu.png?downloads=true&stars=true)](https://nodei.co/npm/tennu/)

----------

## Basic Usage ##

With Tennu, you create an irc client, require your plugins or subscribe to your event listeners, and then connect.

```javascript
var tennu = require('tennu');
var network = require('../config/myNetwork.json'); // See next section
var myClient = tennu.Client(network);
myClient.connect();
```

Before connecting, add listeners to events from irc & users, or load plugins.

```javascript

// Do something when a nick, perhaps yourself, joins a channel
myClient.on('join', function (message) {
    this.say(message.channel, message.actor + " joined!");
});

// Do something when a user emits a command, in this case, join the specified channel.
myClient.on('!join', function (command) {
    this.join(command.args[0]);
});

// Load a plugin.
myClient.initialize(require('./yourModule'));

// Or just use a plugin from tennu_plugins/%f or node_plugins/tennu-%f
myClient.use(['admin', 'last-seen']);

myClient.connect();
```

----------

## Configuration ##

A network configuration object has the following properties:

* server          - IRC server to connect to. _Example:_ _irc.mibbit.net_
* port            - Port to connect to. Defaults to 6667.
* secure          - Use a TLS socket (Throws away the NetSocket)
* ipv6            - Whether you are connecting over ipv6 or not.
* localAddress    - See net.Socket documentation. ;)
* capab           - IRC3 CAP support. (Untested)
* password        - Password for IRC Network (most networks do not have a password)
* nickname        - Nickname the bot will use. Defaults to "tennubot"
* username        - Username the bot will use. Defaults to "user"
* realname        - Realname for the bot. Defaults to "tennu v0.3"
* auth-password   - Password for identifying to services.
* nickserv        - Nickname of nickserv service. Defaults to "nickserv".
* command-trigger - Command character to trigger commands with. By default, '!'.
* channels        - Array of channels to autojoin. _Example:_ ["#help", "#tennu"]
* plugins         - An array of plugin names that the bot requires.
* disable-help    - Disables the built-in help plugin.

Other plugins may add additional properties.

Configuration objects are JSON encodable.

## Dependency Management ##

The second (optional) parameter to tennu.Client is an object of factories to
replace the factories that the Client uses by default.

* NetSocket
* IrcSocket
* MessageHandler
* CommandHandler
* Plugins
* BiSubscriber
* Logger

These functions will always be called as constructors (with `new`).

### Logging ###

Of these, the only one you will probably care about is Logger. The object
returned by the Logger function must implement the following methods:

`debug, info, notice, warn, error, crit, alert, emerg`

Base Tennu will only use debug through error, but other plugins and event
emitters may use crit through emerg.

-------------

## Event Handling ##

Note: Tennu uses a custom event handler. Listeners are placed at the end of the event queue,
insead of happening right away. Errors are currently logged to console, but otherwise
swallowed.

### Respond Functionality ###

Commands and Messages that have a channel property take a return value. Currently, the
return value must be a string or array that is then said to the channel the message
originated in.

```javascript
// Simple echobot.
tennu.on('privmsg', function (privmsg) {
    return privmsg.message;
});

// Equivalent to:
tennu.on('privmsg', function (privmsg) {
    tennu.say(privmsg.channel, privmsg.message);
});
```

### Subscribing Options ###

Subscribing to events in Tennu is more flexible than most event listeners.

You register a single handler on multiple events at once by separating the events with a space,
for example `.on("x y", fn)` is equivalent to `.on('x', fn); .on('y', fn)`. Furthermore, an object
can be passed, where each key is passed as the first parameter and its value, the second.

```javascript
// Examples

on("irc_event", listener)
on("!user-command", listener)
on("join quit", listener)
on({
    "part": part_listener,
    "join": join_listener,
    "!hi !bye": talk_listener
})
```

You can also unsubscribe non-once events with `off`, taking the same parameters as `on`.

### Listener Parameters ###

Listeners are passed either a message or command object.

#### Message ####

Messages are passed by irc events.

Messages are immutable, as are their args. Make sure to copy the args array before trying to manipulate it.

All messages have the following fields:

* receiver   - Receiver of the message. A reference to the Client object.
* prefix     - The prefix is either a hostmask of the format "nickname!username@hostname", or the server you are connected to.
* command    - Message command type. For example, 'privmsg' or 'nick'.
* params     - Array of sent parameters.
* tags       - IRC3 tags sent with message.

Some messages have extended information. See
[Message Properties](http://tennu.github.io/documentation/api/message-properties).

#### Command ####

Commands are passed for user commands.

Commands are an extension of Messages with the command type of 'privmsg'.
They have all properties, plus the following properties:

* args       - Array of words after the command name.
* command    - The command name.

For example, a command of "!do-it ARG1 ARG2" will have args be ["ARG1", "ARG2"] and command be 'do-it'.

--------

## Actions ##

All of the following are methods on Tennu that can be used once connected.

These methods are also available on the client's 'out' property.
In Tennu 0.9.0, the 'out' property will go away, and the 'actions' plugin
will export these methods.

### say(channel, message) ###

* channel is either a channel ("#chan") or a user ("nick").
* message is either a string or array of strings. Given an array, say each
individual element on its own line.

Has the bot say the message(s) to the specific channel/user.

```javascript
/* Output (IRC)
(botnick) This is a message!
*/
tennu.say('#example', "This is a message!");

/* Output (IRC)
(botnick) Hi there.
(botnick) Bye there.
*/
tennu.say('#example', ["Hi there.", "Bye there."]);
```

### act(channel, message) ###

As per say, but as an action (/me)

```javascript
/* Output (IRC)
botnick does something!
*/
tennu.act('#example', "does something!");
```

### ctcp(channel, type, message) ###

```javascript
tennu.ctcp('havvy', 'ping', 'PINGMESSAGE');
```

### nick(newNick) ###

Change the bots nickname.

### join(channel) ###

Joins the specified channel.

```javascript
tennu.join("#tennu");
tennu.join("#keyed-channel channel-key");
tennu.join("#chan1,#chan2");
tennu.join("0"); // Part all channels.
```

### part(channel, reason) ###

Parts the specified channel with the given reason.

### quit(reason) ###

Quits the server with the given reason.

### whois(users, server) ###

Server is optional, and you'll probably not need it. Look at RFC 1459 for
what benefit it gives you.

users is either a string or an array of strings.

### userhost(users) ###

Retrieves the userhost of the user(s).

### raw(message) ###

For actions that are lacking a command, you can use raw to perform them.
You must either pass an array of arguments (and the multiword argument must
be in a single index without the colon) or pass the full string.

If you find yourself using raw(), please file an issue.

### rawf(format, args...) ###

[0.7.1]

As raw(message), but the arguments are passed through util.format() first.

--------

## Plugin System ##

Tennu has its own plugin system.
You can read about it at https://github.com/havvy/tennu-plugins/.

You may access the plugin system's methods via the Client.plugins property
or by using one of the following methods:

* client.require()
* client.getModule()
* client.getRole()
* client.use()
* client.initializePlugin()
* client.isPluginInitializable()

### Creating Your Own Plugins ###

See [Creating Your Own Plugins](https://github.com/Havvy/tennu/blob/master/doc/creating-plugins.md).

See [Getting Started](http://tennu.github.io/documentation/getting-started).

### Built-In Modules ###

Only the help plugin is currently fully implemented.

#### help ####

[0.6.0+]

Sets the command `!help`.

See [Help Module Documentation](https://tennu.github.io/plugins/help).

[0.8.2+]

If you don't want this functionality, set `disable-help` to `true` in your configuration object.

#### channels ####

Unimplemented.

#### users ####

[0.7.3+]

This plugin has a single method exported: isIdentifedAs(nickname, nickname_identified, callback)

See [User Module Documentation](https://tennu.github.io/plugins/user).

#### server ####

[0.9.x+]

Information about the server. For now, the only thing this plugin offers is a
capabilities map listing the information from the 005 raw numeric.

See [Server Plugin Documentation](https://tennu.github.io/plugins/server).

```javascript

var server = tennu.use("server");
console.log(util.inspect(server.capabilities));
```

## Command Line Utility

Install Tennu globally, and you'll gain access to the `tennu` command line tool.

```bash
> pwd
/home/you/your-tennubot
> ls
node_plugins/ tennu_plugins/ config.json
> tennu config.json
```

The tennu command takes two optional argument, -v (--verbose) and -d (--debug),
for adding a Logger that logs to the console (info level and above without -d).

You can also use the tennu command inside your npm scripts when Tennu is installed
locally, and if you are distributing the bot, this is a better option.

## Contributing ##

### Tests ###

```
npm test
```

Between all projects (tennu, tennu-plugins, irc-socket, after-events),
there are over 100 tests, but more are always appreciated, especially
if they are failing with an actual bug. ;)

### Building Files ###

Tennu is written using the Sweet.js dialect of JavaScript. If you've never
used it, it is just JavaScript with macros.

To turn the sweet.js files into javascript files, run `fez.js`.

Place sweet.js files in either `src/` or `test-src/`.

## See Also ##

* [Tennu's Website](https://tennu.github.io/)
* [IRC Specifications and other helpful tables](https://www.alien.net.au/irc/)
* [IRC Wiki](http://www.irc-wiki.org/)


doc/changelog.txt
Changelog

0.1.0:
    * Rewrite of original nrc module.

0.2.0:
    * Module system added.

0.3.0:
    * Removal of the Protocols dependency. There's no logging at the moment.
    * Modules are moved to their own object.
    * BiSubscriber handlers the subscribing of events to either the
      IRCMessageEmitter or the IRCCommandEmitter.
    * Default NRC Modules are disabled (until the next version)

0.4.0: The Module update.
    * Event listeners are no longer bound to 'nrc'.
    * Message and Command objects use 'args' instead of 'params/parameters'.
      This follows the same naming convention that node-irc uses.
    * Default NRC Modules are refactored so that they return a function that
      returns a module.
    * ModuleSubscriber.requires() interface contract changed so that it takes
      a one-arg constructor (that arg being the NRC object for the server the
      module will be on) that returns a module.
    * Module contract gains additional optional field for dependencies:
        + dependencies(undefined U ModuleExports): List[String] U undefined
    * Only the server module works.

0.4.1:
    * Added a new test for Message Handlers.

0.4.2: (Unreleased)
    * Renamed classes.
    * Added an IrcOutputSocket class. NRC absorbs it of course.
    * Added everything to the index.
    * Got Dependency Injection for all non-structure objects.
    * Added `userhost` method to IrcOutputSocket/NRC.

0.4.3:
    * Nrc class renamed to Client.
    * `whois` method added to IrcOutputSocket/Client
    * Moved startup stuff from IrcMessageHandler to Client.
    ** Removed config parameter from its constructor. :)
    * Testing of 'say', 'part', and 'quit' of the IrcOutputSocket.
    ** Feel free to add more tests!

0.4.4:
    * Made dependency injection argument of Client optional.
    * Started the ChunkedMessageParser. Not working.

0.5.0: The Tennu update.
    * Moved module system to its own project.
    * Module system completely revamped.
    * Module loading via name and tennu_modules directory-tree climb.
    * Modules no longer have internal 'name' property.
    * Moved IRC socket to its own project.
    * Got the project building again.
    * Renamed from NRC to Tennu

0.5.1:
    * Fixed OutputSocket.ctcp and OutputSocket.act

0.5.2:
    * Require 1.0.1 of simple-irc-socket for logging.

0.5.3:
    * Hotfix: this._name() doesn't exist in CommandParser

0.6.0: The irc-message update.
    * BACKWARDS INCOMPATIBLE! Changed property names for Message and Command
      to align with IRC specification and the irc-message module.
      See updating.txt for more information.
    * BACKWARDS INCOMPATIBLE! Messages with command type of "nick" have the "newNick" property chaned to "new"
    * BACKWARDS INCOMPATIBLE! Removed the extension for the 353 numeric. It wasn't actually doing anything useful.
      It'll be readded in the future, but actually do some useful parsing. Use .params[0] to get it..
    * Help Module reimplemented. See "help.md" for more information.
    * "Respond" functionality. The return value for listeners that have a channel is said to the channel. No more doing tennu.say(message.channel, what_you_want_to_say). Just do `return what_you_want_to_say;`! In the future, we'll allow promises too.
    * Add .out and .events to Client as aliases to the OutputSocket and BiSubscriber.
    * Messages with the command type of "nick" have the property "old" added, which is an alias for the "nickname" property.
    * Commands have a [[prototype]] of the Message that they are parsed from.
    * Message have a tags property. (See: http://ircv3.org/specification/message-tags-3.2 and https://github.com/expr/irc-message)
    * Removed all default logging.

    * Use irc-message as the base Message parser.
    * simple-irc-socket got renamed to irc-socket, plus is more functional.
    * Replaced mout with lodash.
    * Removed Grunt.js

    * Bugfix: Multiple spaces between words will not mess up commands and message parsing. Thanks expr for reporting.
    * Bugfix: Arbitrary events won't fire when passed incomplete messages from server (fixed in irc-socket)

    * Bug: On command, the client will echo the raw message to the channel.

0.6.1:
    * Bugfix: Commands won't echo the raw message back to the originating channel.
    * Bugfix: Unfreeze Messages parsed by MessageParser.
    * Commands return values are echoed to the originating channel (if they are strings).


lib/client.js
const lodash = require('lodash');
const defaultFactoryConfiguration = {
        'NetSocket': require('net').Socket,
        'IrcSocket': require('irc-socket'),
        'IrcOutputSocket': require('./output-socket.js'),
        'MessageHandler': require('./message-handler.js'),
        'CommandHandler': require('./command-handler.js'),
        'Plugins': require('tennu-plugins'),
        'BiSubscriber': require('./bisubscriber.js'),
        'Logger': require('./null-logger.js'),
        'NicknameTracker': require('./nickname-tracker.js')
    };
const defaultClientConfiguration = {
        'server': undefined,
        'port': 6667,
        'ipv6': undefined,
        'localAddress': undefined,
        'secure': false,
        'password': undefined,
        'capab': false,
        'nickname': 'tennubot',
        'username': 'tennu',
        'realname': 'tennu 0.9.x',
        'channels': [],
        'nickserv': 'nickserv',
        'auth-password': undefined,
        'plugins': [],
        'command-trigger': '!',
        'disable-help': false
    };
/** Fields
 * _config
 * _socket
 * _logger
 * out      (_outputSocket)
 * events   (_subscriber)
 * plugins  (_plugins)
 * nickname (_nickname)
 */
const Client = function (config, dependencies) {
    if (config.nick || config.user) {
        throw new Error('Please use \'nickname\' and \'username\' instead of \'nick\' and \'user\' in your configuration.');
    }
    const client = Object.create(Client.prototype);
    // Parse the configuration object. Make it immutable.
    client._config = config = Object.freeze(lodash.defaults({}, config, defaultClientConfiguration));
    di = lodash.defaults({}, dependencies || {}, defaultFactoryConfiguration);
    // Create a logger.
    // Default logger is a bunch of NOOPs.
    client._logger = new di.Logger();
    // The socket reads and sends messages from/to the IRC server.
    client._socket = new di.IrcSocket(config, di.NetSocket);
    // Create the listener to the socket.
    // This listener will parse the raw messages of the socket, and
    // emits specific events to listen to.
    client._messageHandler = new di.MessageHandler(client, client._logger, client._socket);
    // Create the object that tracks the nickname of the client.
    // Because this object is a function, it is expected that
    // the factory function does not use `this`.
    client._nickname = di.NicknameTracker(config.nickname, client._messageHandler);
    // The output socket wraps the `raw` method of the client._socket.
    client._outputSocket = new di.IrcOutputSocket(client._socket, client._messageHandler, client._nickname, client._logger);
    // Create the listener to private messages from the IRCMessageEmitter
    // The commander will parse these private messages for commands, and
    // emit those commands, also parsed.
    const commandHandler = new di.CommandHandler(config, client, client._nickname, client._logger);
    // The subscriber handles event subscriptions to the Client object,
    // determining whether they should be handled by the IrcMessageEmitter
    // or the Command Handler.
    client._subscriber = new di.BiSubscriber(client._messageHandler, commandHandler);
    client._subscriber.on('privmsg', function (privmsg) {
        commandHandler.parse(privmsg);
    });
    // And finally, the module system.
    client._plugins = new di.Plugins('tennu', client);
    client._plugins.addHook('handlers', function (module, handlers) {
        client._subscriber.on(handlers);
    });
    client.note('Tennu', 'Loading default plugins');
    client._plugins.use([
        'server',
        'help',
        'user',
        'channel',
        'startup'
    ], __dirname);
    client.note('Tennu', 'Loading your plugins');
    client._plugins.use(config.plugins || [], process.cwd());
    client.out = client._outputSocket;
    client.events = client._subscriber;
    client.plugins = client._plugins;
    client.nickname = client._nickname;
    client.connected = false;
    client.note('Tennu', 'Client created.');
    return client;
};
Client.prototype.config = function (param) {
    return this._config[param];
};
// implements Runnable ;)
const connect = function () {
    this.debug('Tennu', 'Trying to connect.');
    if (this.connected) {
        this.warn('Tennu', 'Attempted to start Tennu Client that already is connected.');
        return;
    }
    this._socket.connect();
    this.connected = true;
    this.debug('Tennu', 'Connected');
    return this;
};
Client.prototype.connect = connect;
Client.prototype.start = connect;
const disconnect = function () {
    this.debug('Tennu', 'Trying to disconnect.');
    if (!this.connected) {
        this.warn('Tennu', 'Attempted to end Tennu Client that already is not connected.');
        return;
    }
    this._socket.end();
    this.connected = false;
    this.debug('Tennu', 'Disconnected');
    return this;
};
Client.prototype.disconnect = disconnect;
Client.prototype.end = disconnect;
Client.prototype.say = function () {
    this._outputSocket.say.apply(this._outputSocket, arguments);
    return this;
};
Client.prototype.act = function () {
    this._outputSocket.act.apply(this._outputSocket, arguments);
    return this;
};
Client.prototype.ctcp = function () {
    this._outputSocket.ctcp.apply(this._outputSocket, arguments);
    return this;
};
Client.prototype.part = function () {
    this._outputSocket.part.apply(this._outputSocket, arguments);
    return this;
};
Client.prototype.quit = function () {
    this._outputSocket.quit.apply(this._outputSocket, arguments);
    return this;
};
Client.prototype.join = function () {
    this._outputSocket.join.apply(this._outputSocket, arguments);
    return this;
};
Client.prototype.userhost = function () {
    this._outputSocket.userhost.apply(this._outputSocket, arguments);
    return this;
};
Client.prototype.whois = function () {
    this._outputSocket.whois.apply(this._outputSocket, arguments);
    return this;
};
Client.prototype.nick = function () {
    this._outputSocket.nick.apply(this._outputSocket, arguments);
    return this;
};
Client.prototype.mode = function () {
    this._outputSocket.mode.apply(this._outputSocket, arguments);
    return this;
};
Client.prototype.raw = function () {
    this._outputSocket.raw.apply(this._outputSocket, arguments);
    return this;
};
Client.prototype.rawf = function () {
    this._outputSocket.rawf.apply(this._outputSocket, arguments);
    return this;
};
Client.prototype.on = function () {
    this._subscriber.on.apply(this._subscriber, arguments);
    return this;
};
Client.prototype.once = function () {
    this._subscriber.once.apply(this._subscriber, arguments);
    return this;
};
Client.prototype.off = function () {
    this._subscriber.off.apply(this._subscriber, arguments);
    return this;
};
Client.prototype.use = function () {
    this._plugins.use.apply(this._plugins, arguments);
    return this;
};
Client.prototype.getModule = function () {
    this._plugins.moduleExports.apply(this._plugins, arguments);
    return this;
};
Client.prototype.getRole = function () {
    this._plugins.roleExports.apply(this._plugins, arguments);
    return this;
};
Client.prototype.initializePlugin = function () {
    this._plugins.initialize.apply(this._plugins, arguments);
    return this;
};
Client.prototype.isPluginInitializable = function () {
    this._plugins.isInitializable.apply(this._plugins, arguments);
    return this;
};
Client.prototype.addHook = function () {
    this._plugins.addHook.apply(this._plugins, arguments);
    return this;
};
Client.prototype.debug = function () {
    this._logger.debug.apply(this._logger, arguments);
    return this;
};
Client.prototype.info = function () {
    this._logger.info.apply(this._logger, arguments);
    return this;
};
Client.prototype.notice = function () {
    this._logger.notice.apply(this._logger, arguments);
    return this;
};
Client.prototype.note = function () {
    this._logger.notice.apply(this._logger, arguments);
    return this;
};
Client.prototype.warn = function () {
    this._logger.warn.apply(this._logger, arguments);
    return this;
};
Client.prototype.error = function () {
    this._logger.error.apply(this._logger, arguments);
    return this;
};
Client.prototype.crit = function () {
    this._logger.crit.apply(this._logger, arguments);
    return this;
};
Client.prototype.alert = function () {
    this._logger.alert.apply(this._logger, arguments);
    return this;
};
Client.prototype.emerg = function () {
    this._logger.emerg.apply(this._logger, arguments);
    return this;
};
Client.prototype.log = function (level) {
    const args = Array.prototype.slice.call(arguments, 1);
    this[level].apply(this, args);
};
// Export the factory.
module.exports = Client;